File: src/main/java/com/senderman/lastkatkabot/command/user/BncTop.java
package com.senderman.lastkatkabot.command.user;

import com.annimon.tgbotsmodule.api.methods.Methods;
import com.annimon.tgbotsmodule.commands.context.MessageContext;
import com.annimon.tgbotsmodule.services.CommonAbsSender;
import com.senderman.lastkatkabot.command.CommandExecutor;
import com.senderman.lastkatkabot.dbservice.UserStatsService;
import com.senderman.lastkatkabot.util.Html;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.ChatMember;
import org.telegram.telegrambots.meta.api.objects.User;

import java.util.Optional;
import java.util.function.Function;

@Component
public class BncTop implements CommandExecutor {

    private final UserStatsService users;

    public BncTop(UserStatsService users) {
        this.users = users;
    }

    @Override
    public String getTrigger() {
        return "/bnctop";
    }

    @Override
    public String getDescription() {
        return "топ игроков в Быки и Коровы";
    }

    @Override
    public void execute(MessageContext ctx) {

        int counter = 0;
        var top = new StringBuilder("<b>Топ-10 задротов в bnc:</b>\n\n");
        var topUsers = users.findTop10BncPlayers();
        for (var user : topUsers) {
            top.append(++counter)
                    .append(": ")
                    .append(formatUser(user.getUserId(), user.getBncScore(), ctx.message().isUserMessage(), ctx.sender))
                    .append("\n");
        }

        ctx.reply(top.toString()).callAsync(ctx.sender);
    }

    private String formatUser(long userId, int score, boolean printLink, CommonAbsSender telegram) {
        Function<User, String> userPrinter = printLink ? Html::getUserLink : u -> Html.htmlSafe(u.getFirstName());

        String user = Optional.ofNullable(Methods.getChatMember(userId, userId).call(telegram))
                .map(ChatMember::getUser)
                .map(userPrinter)
                .orElse("Без имени");

        return user + " (" + score + ")";
    }
}

File: src/main/kotlin/com/senderman/lastkatkabot/bnc/commands/BNCTop.kt
package com.senderman.lastkatkabot.bnc.commands

import com.annimon.tgbotsmodule.api.methods.Methods
import com.senderman.lastkatkabot.LastkatkaBotHandler
import com.senderman.lastkatkabot.Services
import com.senderman.lastkatkabot.bnc.BNCPlayer
import com.senderman.neblib.CommandExecutor
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

class BNCTop(private val handler: LastkatkaBotHandler) : CommandExecutor {
    override val command: String
        get() = "/top"
    override val desc: String
        get() = "топ игроков в Быки и Коровы"

    override fun execute(message: Message) {
        val chatId = message.chatId
        val top = Services.db.getTop()
        val text = StringBuilder("<b>Топ-10 задротов в bnc:</b>\n\n")
        var counter = 1
        for ((playerId, score) in top) {
            val name = try {
                Methods.getChatMember(playerId.toLong(), playerId).call(handler).user.firstName
            } catch (e: TelegramApiException) {
                "Без имени"
            }
            val player = BNCPlayer(playerId, name, score)
            text.append(counter).append(": ")
            if (message.isUserMessage) text.append(player.link) else text.append(player.name)
            text.append(" (${player.score})\n")
            counter++
        }
        handler.sendMessage(chatId, text.toString())
    }
}